doc/SAI-Proposal-15-UOM.md

```diff
@@ -0,0 +1,229 @@
+
+### Hidden SAI Objects after switch initialization
+
+By default, some SAI objects are created during switch initialization, such as port, default VLAN. In order for the SAI view to capture the full ASIC states, such objects must be queryable through SAI API. For this purpose, we will be using sai\_get\_object\_key which was purpose in the warmboot proposal for this purpose.
+
+## Examples
+
+## Spec
+
+Following changes to current SAI spec are proposed.
+
+- Add VLAN member object
+- Add ECMP group member object
+- Change TRAP ID to unified object ID, add Create/Remove API for trap id
+- Add Port Create/Remove API to handle port breakout/break-in 
+- Add Queue Create/Remove API for dynamic queue creation/deletion 
+
+### VLAN member
+    
+Add VLAN member object and create/remove/get/set API. 
+
+    typedef enum _sai_vlan_member_attr_t {
+ 
+        /** READ_WRITE */
+ 
+        /** VLAN ID [sai_vlan_t] (MANDATORY_ON_CREATE|CREATE_ONLY) */
+        SAI_VLAN_MEMBER_ATTR_VLAN_ID,
+ 
+        /** logical port ID [sai_object_id_t] (MANDATORY_ON_CREATE|CREATE_ONLY) */
+        SAI_VLAN_MEMBER_ATTR_PORT_ID,
+ 
+        /** VLAN tagging mode [sai_vlan_tagging_mode_t] (CREATE_AND_SET)
+         * (default to SAI_VLAN_PORT_UNTAGGED) */
+        SAI_VLAN_MEMBER_ATTR_TAGGING_MODE,
+ 
+        /** custom range base value */
+        SAI_VLAN_MEMBER_ATTR_CUSTOM_RANGE_BASE  = 0x10000000
+ 
+    } sai_vlan_member_attr_t;
+
+    typedef sai_status_t (*sai_create_vlan_member_fn)(
+        _Out_ sai_object_id_t* vlan_member_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list
+        );
+
+    typedef sai_status_t (*sai_remove_vlan_member_fn)(
+        _In_ sai_object_id_t vlan_member_id
+       );
+
+    typedef sai_status_t (*sai_set_vlan_member_attribute_fn)(
+       _In_ sai_object_id_t vlan_member_id,
+       _In_ const sai_attribute_t *attr
+       );
+ 
+    typedef sai_status_t (*sai_get_vlan_member_attribute_fn)(
+       _In_ sai_object_id_t vlan_member_id,
+       _In_ const uint32_t attr_count,
+       _Inout_ sai_attribute_t *attr_list
+       );
+
+In the VLAN object, add SAI_VLAN_ATTR_MEMBER_LIST to query the vlan members. remove SAI_VLAN_ATTR_PORT_LIST. 
+ 

```

Comment:
Consider retaining it as a read only attribute. Will be helpful for the host adapter to get the list if it needs.